python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -516,7 +516,7 @@ def pipe(self, func, *args, **kwargs):
         """
         return cudf.core.common.pipe(self, func, *args, **kwargs)
 
-    def apply(self, function):
+    def apply(self, function, *args):
```

Comment:
Is there a reason we can't support `**kwargs` in the same way?

Reply:
Good question. I considered this, and the reason I did not add it was really roadmap considerations. I know we want to rethink how this API works in general, and if we start supporting this now, we could back ourselves into a corner where users are reliant upon it and we're limited in the avenues we can pursue to replace this pipeline without complicating things at the outset. The same thing makes me a little uneasy about even adding `*args`, but at least there's a direct feature request for that and precedent for it elsewhere (`DataFrame.apply` and `Series.apply` both support `args=`). 

As you have observed, it would probably not take much work to add so I'm happy to add it if there's strong motivation to do so.

Reply:
The motivation is to make this API align with pandas. https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.apply.html#pandas.core.groupby.GroupBy.apply

What are we wanting to rethink? The public API, or the internal implementation? I would say that `*args` and `**kwargs` go together for this feature and we should implement both or neither -- not just `*args`.

Reply:
Similarly, we should enable `**kwargs` for `Series.apply` and `DataFrame.apply` if the numba machinery is no more complicated than it is here. Parameters-by-keyword and parameters-by-position serve the same purpose and I don't see a fundamental distinction in complexity between them.

Reply:
It's the internal implementation that I expect to change, since this implementation does a loop over chunks which is what makes it potentially slow. Right now this API isn't backed by Numba at all, which is why we can pass through `args` and theoretically `kwargs` rather easily. 

`kwargs` in a numba implementation is not trivial. For context, numba [does not support kwargs](https://numba.pydata.org/numba-doc/dev/reference/pysupported.html) at all in compiled functions. This is why we don't have it in the other `apply` APIs. Since we could end up exploring some kind of JIT compiled implementation of this API as well, it might be closing the door on some solutions for us that could otherwise cover a broad swath of use cases.

> I would say that `*args` and `**kwargs` go together for this feature and we should implement both or neither -- not just `*args`.

I empathize with the feeling of asymmetry that comes along with having `*args` and not `**kwargs`, but not with the notion that we should not have `*args` unless we can also have `**kwargs`. For reasons similar to what you noted about there being little fundamental distinction, supporting `*args` unlocks a space of functions including many logical equivalents to `**kwargs` functions while still allowing us to tiptoe around this API a bit. 

What do you think, @bdice ?

Reply:
Okay, that's the context I was missing. 😄 Let's move forward with this as-is, then. I'll review the rest of the PR.